### PR TITLE
[Model Runner V2][Spec Decode] Fix stale values in idx_mapping from CG num reqs padding

### PR DESCRIPTION
**Pull request by TheEpicDolphin**
# Context
During full cudagraphs, num_reqs can be padded up to the next capture size. Currently, the speculators can run `sample_draft` as part of a cudagraph, using the padded num_reqs first dimension size for `idx_mapping`. This is problematic because stale values in the `idx_mapping` within the padding region can be used during gumbel sampling to write to the `draft_logits` tensor. The consequence is that when running with the `probabilistic` draft sampling method during full cudagraphs, we can get incorrect/gibberish outputs.

# This PR
Pads the `idx_mapping` past `num_reqs` with -1. During gumbel sampling, this value is now used to detect invalid requests, which are skipped.

# Testing
## Server
```
VLLM_USE_V2_MODEL_RUNNER=1 vllm serve deepseek-ai/DeepSeek-V4-Flash \
    -tp 4 \
    --block-size 256 \
    --max-num-seqs 64 \
    --kv-cache-dtype fp8 \
    --tokenizer-mode deepseek_v4 \
    --tool-call-parser deepseek_v4 \
    --reasoning-parser deepseek_v4 \
    --moe-backend deep_gemm \
    --no-enable-prefix-caching \
    --kernel-config.enable_flashinfer_autotune=False \
    --speculative-config '{"method": "mtp", "num_speculative_tokens": 3, "rejection_sample_method": "standard", "draft_sample_method":"probabilistic"}'
```

## Bench
```
vllm-bench \
    --dataset-name hf \
    --dataset-path philschmid/mt-bench \
    --backend openai-chat \
    --endpoint /v1/chat/completions \
    --request-rate inf \
    --max-concurrency 32 \
    --num-prompts 50 \
    --temperature 1.0 \
    --extra-body '{"chat_template_kwargs":{"thinking":false}}' \
    --save-result \
    --save-detailed \
    --result-dir ~/dev
```

## Outputs
**Before**
8 out of the 50 responses have issues like token duplication, word fusion/salad, etc. Two examples:
```
"Here is a detailed 3-day lesson plan integrating Drama\n\n**Mime** techniques techniques to explore the complexities of the Opium Wars. This plan emphasizes empathy, non-verbal communication, and the clash of cultural perspectives.\n\n**Lesson Plan: The Opium Wars – A Clash of Wills (Through Mime & Tableau)**\n\n**Grade Level:** 9-10 (Ages 14-16)\n**Topic:** The Opium Wars (1839-1842, 1856-1860)\n**Subject:** History (with integratedated Drama **Theatre (Drama, History**&ral\n**Categories.”#"

"I'll clarify what a \"highest common ancestor\" means in this context. The **highest common ancestor** is the deepestThe common ancestorTryI'll clarifyHighest common ancestor\" means the deepest node,This is actually we need toalkingI nCommon, I clarificationTheThe **deepest common ancestor** *(Shortseg: The HighestO... I need of two nodes - f = .\n\nGiven a node deepest This indicates the common ancestorAncestor immediately deeper, I need to ......\n\nThe **highest common ancestor** (HCA) of two nodes is the deepestCommon the deepestI'llThe common ancestor, I"
```

**After**
All 50 responses are correct and coherent. The two incorrect responses from before now look like:
```
"Here is a comprehensive 3-day lesson plan for 9th-10th grade history students, integrating drama, mime, and theater techniques to explore the Opium Wars.\n\n---\n\n### Lesson Plan: The Opium Wars – Through the Lens of Drama\n\n**Subject:** History (World History / Asian Studies)\n**Grade Level:** 9-10\n**Topic:** The Opium Wars (1839-1842 & 1856-1860) between China and Britain\n**Total Duration:** Three 45-minute class periods\n\n**Essential Question:** How did different perspectives, motivations, and power imbalances shape the events and consequences"

"I'll create a function to find the Highest Common Ancestor (HCA) of two nodes in a binary tree. Note that \"Highest Common Ancestor\" typically refers to the ancestor that is closest to the root (highest in the tree), which is different from the Lowest Common Ancestor (LCA).\n\n```python\nclass TreeNode:\n    def __init__(self, val=0, left=None, right=None):\n        self.val = val\n        self.left = left\n        self.right = right\n\ndef highest_common_ancestor(root: TreeNode, node1: TreeNode, node2: TreeNode) -> TreeNode:\n"
```

### CODE REVIEW

File: vllm/v1/worker/gpu/sample/gumbel.py [hunk 1]
```diff
@@ -100,7 +100,10 @@ def gumbel_block_argmax(
     PER_TOKEN_COL: tl.constexpr = False,
 ):
     req_state_idx = tl.load(expanded_idx_mapping_ptr + token_idx).to(tl.int64)
```
Comment (benchislett) [inline vllm/v1/worker/gpu/sample/gumbel.py:102]: Naive Q: Is there any risk of the underlying index tensor being unsigned ints? And -1 underflowing?

File: vllm/v1/worker/gpu/sample/gumbel.py [hunk 2]
```diff
@@ -131,7 +134,7 @@ def gumbel_block_argmax(
         logits = logits.to(tl.float64)
     if temp != 0.0:
         # Calculate the seed for gumbel noise.
-        seed = tl.load(seeds_ptr + req_state_idx)
+        seed = tl.load(seeds_ptr + req_state_idx, mask=is_valid_req, other=0)
```
Comment (benchislett) [inline vllm/v1/worker/gpu/sample/gumbel.py:137]: I think this is unreachable because temp is always 0 when the mask indicates inactive threads? Good to have for sanity though, just pointing it out

**Review by benchislett — APPROVED** [review]
LGTM

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py [hunk 3]
```diff
@@ -366,7 +366,7 @@ def propose(
             self.query_cudagraph_manager.run_fullgraph(batch_desc)
         else:
             self._generate_draft(
-                num_reqs_padded,
+                num_reqs,
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dflash/speculator.py:369]: Why?

Comment (TheEpicDolphin) [reply vllm/v1/worker/gpu/spec_decode/dflash/speculator.py:369]: i noticed that we are processing the padded requests whenever we fall back to eager mode, potentially adding some extra overhead. Not crucial for this fix

File: vllm/v1/worker/gpu/sample/gumbel.py (hunk 1 above)
Comment (TheEpicDolphin) [reply vllm/v1/worker/gpu/sample/gumbel.py:102]: The codebase consistently uses signed int for idx_mapping, so this should be relatively safe
